src/core/entity.js

```diff
@@ -93,46 +113,47 @@ class Entity {
     this.entities.push(entity);
   }
 
-  get_matrix(out) {
+  get_view_matrix(out) {
     const look_at_vect = [];
     math.vec3.add(look_at_vect, this.position, this.forward);
     math.mat4.look_at(out, this.position, look_at_vect, this.up);
     return out;
   }
 
-  rotate_rl(rad) {
-    const rightMatrix = math.mat4.create();
-    math.mat4.rotate(rightMatrix, rightMatrix, rad, this.up);
-    math.vec3.transform_mat4(this.forward, this.forward, rightMatrix);
-    this.realign();
+  rotate_along(vec, rad) {
+    // Negate the origin vector so that we can go back.
+    const reverse_origin = this.origin.map(i => -i);
+    const rotation_matrix = math.mat4.create();
+    math.mat4.from_rotation(rotation_matrix, rad, vec);
+    math.mat4.translate(this.local_matrix, this.local_matrix, this.origin);
+    math.mat4.multiply(this.local_matrix, this.local_matrix, rotation_matrix);
+    math.mat4.translate(this.local_matrix, this.local_matrix, reverse_origin);
```

Comment:
The order of the translations here is reversed compared to the current v1.0.  I didn't get to test this yet, so this might be wrong.

Reply:
I also think that we should remove `origin` for now: the same effect can be achieved with parents.  It's unclear to me right now if children's origin should be used for rotating them if the parent is rotated.

Reply:
I don't fully understand your concerns here, maybe it's because `origin` isn't the most fortunate name for this attribute.
Lemme visualise my idea behind it. 

### Default origin
```js
cube.origin = [0, 0, 0];
game.add_frame_listener('do_stuff', delta => {
  cube.rotate_rl(16/1000);
});
```
the cube's rotating in place.
![rotate-no-origin](https://user-images.githubusercontent.com/347657/29732347-85228540-89e8-11e7-8cca-fb58b49e1bc7.gif)

### Non-default origin
```js
cube.origin = [ -2, 0, 0];
game.add_frame_listener('do_stuff', delta => {
  cube.rotate_rl(16/1000);
});
```
The entity is still rotating, but the point of rotation (the `origin`) is displaced:
![rotate-origin](https://user-images.githubusercontent.com/347657/29732411-ca1f5614-89e8-11e7-8603-a7a66b7289f0.gif)

Of course you **can** achieve the same thing with a parent entity, but it increases number of elements in scene, number of computations per frame, and is probably harder to use than just changing yet another attribute on the class.